src/MSBuild/TerminalLogger/TerminalLogger.cs

```diff
@@ -380,6 +390,30 @@ private void ProjectStarted(object sender, ProjectStartedEventArgs e)
                 int nodeIndex = NodeIndexForContext(buildEventContext);
                 _nodes[nodeIndex] = new NodeStatus(e.ProjectFile!, null, "Restore", _projects[c].Stopwatch);
             }
+
+            TryDetectGenerateFullPaths(e, project);
+        }
+    }
+
+    private void TryDetectGenerateFullPaths(ProjectStartedEventArgs e, Project project)
+    {
+        if (e.GlobalProperties is not null
+            && e.GlobalProperties.TryGetValue("GenerateFullPaths", out string? generateFullPaths)
+            && bool.TryParse(generateFullPaths, out bool generateFullPathsValue))
+        {
+            project.GenerateFullPaths = generateFullPathsValue;
+        }
+        else if (e.Properties is not null)
+        {
+            foreach (DictionaryEntry property in e.Properties)
+            {
+                if (property.Key is "GenerateFullPaths" &&
```

Comment:
Should this be a case-insensitive comparison to match `e.GlobalProperties` which has a case insensitive key comparer?

Reply:
Oh I'm going nuts on the implementation still trying to find something that consistently works with ProjectEvaluationFinishedEventArgs, I wouldn't check any of this as 'final' yet. 

I'm running into a kind of frustrating scenario with this PR switching from ProjectStarted to ProjectEvaluationFinishedEventArgs.

The `Project` items we create are keyed to specific `ProjectContextId`s from the BuildEventContext, but those `ProjectContextId`s are not the same between evaluation and execution. The _projectInstances_ are, though - should we consider keying the `ProjectContext` type by that instead, so that properties/items gotten during evaluation can be kept associated with the _executing_ project?

Reply:
Have you tried matching on `EvaluationId`? It looks like the value passed to `ProjectStarted` matches the one in `ProjectEvaluationFinishedEventArgs`.

Reply:
@ladipro there's something here that's making me pretty concerned.  I put breakpoints on 3 locations in the terminal logger

* where we handle ProjectEvaluationFinishedEventArgs
* where we handle ProjectStartedEventArgs
* where we handle TaskParameterEventArgs

and at no point is there an Id that remains the same across all of them. Here's some data from a recent test run to highlight the problem:

```
Restore evaluation (ProjectEvaluationFinishedEventArgs)
    {Node=1 Submission=0 ProjectContext=-2 ProjectInstance=-1 Eval=4 Target=-1 Task=-1}
Restore execution (ProjectStarted)
    {Node=1 Submission=0 ProjectContext=4 ProjectInstance=2 Eval=4 Target=-1 Task=-1}
_IsProjectRestoreSupported evaluation (ProjectEvaluationFinishedEventArgs)
    {Node=1 Submission=0 ProjectContext=-2 ProjectInstance=-1 Eval=7 Target=-1 Task=-1}
_IsProjectRestoreSupported execution (ProjectStarted)
    {Node=1 Submission=0 ProjectContext=7 ProjectInstance=3 Eval=7 Target=-1 Task=-1}
_GenerateRestoreProjectPathWalk execution (ProjectStarted)
    {Node=1 Submission=0 ProjectContext=10 ProjectInstance=3 Eval=7 Target=-1 Task=-1}
_IsProjectRestoreSupported execution (ProjectStarted)
    {Node=1 Submission=0 ProjectContext=13 ProjectInstance=3 Eval=7 Target=-1 Task=-1}
_GenerateRestoreGraphProjectEntry execution (ProjectStarted)
    {Node=1 Submission=0 ProjectContext=16 ProjectInstance=3 Eval=7 Target=-1 Task=-1}
_GenerateProjectRestoreGraph execution (ProjectStarted)
    {Node=1 Submission=0 ProjectContext=19 ProjectInstance=3 Eval=7 Target=-1 Task=-1}
'actual' evaluation (ProjectEvaluationFinishedEventArgs)
    {Node=1 Submission=1 ProjectContext=-2 ProjectInstance=-1 Eval=10 Target=-1 Task=-1}
'actual' execution (ProjectStarted)
    {Node=1 Submission=1 ProjectContext=22 ProjectInstance=4 Eval=10 Target=-1 Task=-1}
TaskItem emit (TaskParameterEventArgs) context - Why is Eval -1?
    {Node=1 Submission=1 ProjectContext=22 ProjectInstance=4 Eval=-1 Target=135 Task=-1}
```

Each line is the build context at each state - my conclusion is there is no single Id that we can use to keep a shared context across evaluation, execution, and arbitrary task messages. I think we might need to keep/track both ProjectInstance and Eval Ids and be able to look up TerminalLogger's `Project` based on either of those.

Reply:
I figured it out - I made a parallel tracker for data we are interested in from evaluation time (which right now is just TFM) and then use the `BuildMessageContext.Eval` property as the key for that evaluation data. This is working great and should be a good model for us in the future - we can use the new `EvaluationData` structure to hold specific data that's relevant just from an evaluation, and the `Project` structure should contain all data that changes or is computed at target execution time.